Source/DafnyCore/Compilers/CSharp/Compiler-Csharp.cs

```diff
@@ -265,27 +265,57 @@ public class CsharpCompiler : SinglePassCompiler {
 
           var wCtorParams = new ConcreteSyntaxTree();
           wCtorBody = wBody.Format($"public {name}({wCtorParams})").NewBlock();
+          EmitTypeDescriptorsForClass(typeParameters, cls, wTypeFields, wCtorParams, null, wCtorBody);
+        }
+      }
 
-          if (typeParameters != null) {
-            var sep = "";
-            foreach (var ta in TypeArgumentInstantiation.ListFromFormals(typeParameters)) {
-              if (NeedsTypeDescriptor(ta.Formal)) {
-                var fieldName = FormatTypeDescriptorVariable(ta.Formal.GetCompileName(Options));
-                var decl = $"{DafnyTypeDescriptor}<{TypeName(ta.Actual, wTypeFields, ta.Formal.tok)}> {fieldName}";
-                wTypeFields.WriteLine($"private {decl};");
-                if (ta.Formal.Parent == cls) {
-                  wCtorParams.Write($"{sep}{decl}");
-                }
+      return new ClassWriter(this, wBody, wCtorBody);
+    }
 
-                wCtorBody.WriteLine($"this.{fieldName} = {TypeDescriptor(ta.Actual, wCtorBody, ta.Formal.tok)};");
-                sep = ", ";
-              }
+    /// <summary>
+    /// For each type parameter X in "typeParametersForClass" that needs a type descriptor,
+    ///   * Write "protected TypeDescriptor<X> _td_X;" to wTypeFields
+    ///     -- each entry is terminated by a newline
+    ///   * Write "TypeDescriptor<X> _td_X" to wCtorParams
+    ///     -- entries are separated by a comma
+    ///   * Write "_td_X" to wCallArguments
+    ///     -- entries are separated by a comma
+    ///   * Write "this._td_X := _td_X;" to wCtorBody
+    ///     -- each entry is terminated by a newline
+    /// Any of the writer parameters may be null, so long as at least one is non-null.
+    /// The method returns the number type descriptors written.
+    /// </summary>
+    int EmitTypeDescriptorsForClass(List<TypeParameter> typeParametersForClass, TopLevelDecl cls,
```

Comment:
Consider returning values instead of mutating arguments. It moves the work of creating some of these SyntaxTrees from the caller to the callee.

```c#

(int writtenDescriptors, 
 ICanRender typeFields, 
 ICanRender constructorParams,
 ICanRender callArguments,
 ICanRender constructorBody) EmitTypeDescriptorsForClass(
  List<TypeParameter> typeParametersForClass, 
  TopLevelDecl cls,
  List<TypeParameter> alternateTypeParameters = null)
{
  var typeFields = new ConcreteSyntaxTree();
  ..
}

```

Reply:
I changed the in-parameters to out-parameters and reworked the call sites.